src/metric/xentropy_metric.hpp

```diff
@@ -0,0 +1,381 @@
+    		Log::Fatal("[%s]: metric does not tolerate label [#%i] outside [0, 1]", __func__, i);
+    	}
+    }
+    Log::Info("[%s:%s]: labels passed interval [0, 1] check (metric)",  GetName()[0].c_str(), __func__);
+
+    if (weights_ == nullptr) {
+      sum_weights_ = static_cast<double>(num_data_);
+    } else {
+      sum_weights_ = 0.0f;
+      for (data_size_t i = 0; i < num_data; ++i) {
+        sum_weights_ += weights_[i];
+      }
+    }
+    // check weight sum
+    if (sum_weights_ <= 0.0f) {
+      Log::Fatal("[%s]: sum-of-weights = %f is non-positive", __func__, sum_weights_);
+    } else {
+      Log::Info("[%s:%s]: sum-of-weights = %f",  GetName()[0].c_str(), __func__, sum_weights_);
+    }
+  }
+
+  std::vector<double> Eval(const double* score, const ObjectiveFunction* objective) const override {
+    double sum_loss = 0.0f;
+    if (objective == nullptr) {
+      Log::Warning("[%s : %d]: objective nullptr", __FILE__, __LINE__);
+      if (weights_ == nullptr) {
+        #pragma omp parallel for schedule(static) reduction(+:sum_loss)
+        for (data_size_t i = 0; i < num_data_; ++i) {
+          sum_loss += XentLoss(label_[i], score[i]); // NOTE: does not work unless score is a probability
+        }
+      } else {
+        #pragma omp parallel for schedule(static) reduction(+:sum_loss)
+        for (data_size_t i = 0; i < num_data_; ++i) {
+          sum_loss += XentLoss(label_[i], score[i]) * weights_[i]; // NOTE: does not work unless score is a probability
+        }
+      }
+    } else {
+      if (weights_ == nullptr) {
+        #pragma omp parallel for schedule(static) reduction(+:sum_loss)
+        for (data_size_t i = 0; i < num_data_; ++i) {
+          double p = 0;
+          objective->ConvertOutput(&score[i], &p);
+          sum_loss += XentLoss(label_[i], p);
+        }
+      } else {
+        #pragma omp parallel for schedule(static) reduction(+:sum_loss)
+        for (data_size_t i = 0; i < num_data_; ++i) {
+          double p = 0;
+          objective->ConvertOutput(&score[i], &p);
+          sum_loss += XentLoss(label_[i], p) * weights_[i];
+        }
+      }
+    }
+    double loss = sum_loss / sum_weights_;
+    return std::vector<double>(1, loss);
+  }
+
+	const std::vector<std::string>& GetName() const override {
+    return name_;
+  }
+
+  double factor_to_bigger_better() const override {
+    return -1.0f; // negative means smaller loss is better, positive means larger loss is better
+  }
+
+private:
+  /*! \brief Number of data points */
+  data_size_t num_data_;
+  /*! \brief Pointer to label */
+  const float* label_;
+  /*! \brief Pointer to weights */
+  const float* weights_;
+  /*! \brief Sum of weights */
+  double sum_weights_;
+  /*! \brief Name of this metric */
+  std::vector<std::string> name_;
+};
+
+//
+// CrossEntropyMetric1 : "xentropy1" : (optional) weights have a different meaning than for "xentropy"
+//
+class CrossEntropyMetric1 : public Metric {

```

Comment:
Please using an self-explained class name for this `"intensity-weighted" ` metric